### PR TITLE
Build the file list and run linters in parallel

### LINKED ISSUE #5073: JSON linting still very slow on medium size files
### Is there an existing issue for this?

- [X] I have searched the existing issues. Similar issue https://github.com/super-linter/super-linter/issues/2713 and made comment https://github.com/super-linter/super-linter/issues/2713#issuecomment-1870653818

### Current Behavior

I have approximaely 300 mb+ of json files linting in my code base with the biggest files being about 150kb so not huge. This is taking several hours to lint on standard Github runners. A little slow but several hours seems the wrong order of magnitude for JSON linting which should be fairly processor light process.

When debugging is enabled I can see lots of other  linters are running despite only having json linting enabled.

### Expected Behavior

* Settings show can either enable the linters you want to run or disable the ones you don't. When enabling just JSON linting the expectation would be that only that is run, the majority of use cases of JSON linting do not need to check whether correctly validated as ARM file for example.
* Takes a few minutes to run

### Super-Linter version

```markdown
v5.7.2 which as of writing is the latest version
```

### Relevant log output

```shell
....

2023-12-27 21:56:33 [DEBUG]   File:[/path/subpath/xxx.json], File_type:[json], Base_file:[xxx.json]
2023-12-27 21:56:33 [DEBUG]   Checking if /path/subpath/xxx.json is an OpenAPI file...
2023-12-27 21:56:33 [DEBUG]   /path/subpath/xxx.json is NOT an OpenAPI descriptor
2023-12-27 21:56:33 [DEBUG]   Checking if /path/subpath/xxx.json is an ARM file...
2023-12-27 21:56:33 [DEBUG]   Checking if /path/subpath/xxx.json is a Cloud Formation file...
2023-12-27 21:56:33 [DEBUG]   Checking if /path/subpath/xxx.json is a AWS states descriptor...

...
```

### Steps To Reproduce

Github Workflow Config

```
      - name: Lint Code Base
        id: super
        uses: super-linter/super-linter/slim@v5.7.2
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          VALIDATE_ALL_CODEBASE: true 
          VALIDATE_JSON: true
          FILTER_REGEX_INCLUDE: (./some_location\/.*)
          ACTIONS_RUNNER_DEBUG: true
```

### PR DESCRIPTION
**Pull request by ferrarimarco**
# Proposed changes

- Parallelize building the file list
- Parallelize linters
- Enable Gitleaks to run against single files because it's parallelized, so there's no reason to treat it in a special way.

Fix #5073

The tasks I parallelized are, by far, the most expensive tasks. We could also parallelize the configuration of linter config files, but that takes a few seconds only.

To implement these changes, I had to refactor things a bit:

- Don't use a `.gitleaksignore` because Gitleaks doesn't provide a way to ignore this file. This was causing issues when running tests.
- Don't handle special cases to exclude `bad` and `good` tests. Use `FILTER_REGEX_EXCLUDE` and `FILTER_REGEX_INCLUDE` instead.
- Fix an issue with Checkov not linting files if the configuration file included the list of directories to lint.
- Move the logic to run linter tests to a script to reduce duplication in the `Makefile`, and to enable further error checks. I'll reduce duplication even more in a follow up PR.
- Remove the `WORKSPACE_PATH` variable. Use `GITHUB_WORKSPACE` only to reduce the things to consider.
- Use temporary files to hold the list of files to lint, and return codes so that processes can read them. This is necessary to implement a robust inter-process communication. Environment variables are too fragile for this purpose.
- Move linter commands to a dedicated file that processes can source as needed because Bash doesn't currently offer a reliable way to export arrays.
- Save output in JSON files and parse them to print it as needed.
- Export some functions and variables so that subprocesses can use them.
- Move the logic to run "additional installations" to the `RunAdditionalInstalls` function instead of having them scattered around several places.
- Use arrays to store commands instead of an associative array of strings to properly handle escapes and arguments.
- To differentiate between different test cases, we now return: `0` if everything was linted fine, `1` if there were errors, `2` if there were errors, but some linters reported success.

## Readiness checklist

In order to have this pull request merged, complete the following tasks.

### Pull request author tasks

- [x] I included all the needed documentation for this change.
- [x] I provided the necessary tests.
- [x] I squashed all the commits into a single commit.
- [x] I followed the [Conventional Commit v1.0.0 spec](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I wrote the necessary upgrade instructions in the [upgrade guide](../docs/upgrade-guide.md).
- [x] If this pull request is about and existing issue,
  I added the `Fix #ISSUE_NUMBER` label to the description of the pull request.

### Super-linter maintainer tasks

- [x] Label as `breaking` if this change breaks compatibility with the previous released version.
- [x] Label as either: `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`.

### CODE REVIEW

**Review by Hanse00 — COMMENTED** [review]
Given the size and complexity of these changes, I honestly did more of a skim than an in-depth review of the code changes. But nothing immediately stood out to me besides the two comments attached.

File: .gitleaksignore [hunk 1]
```diff
@@ -1,9 +0,0 @@
-.github-personal-access-token:github-fine-grained-pat:1
-/github/workspace/.github-personal-access-token:github-fine-grained-pat:1
-/github/workspace/test/linters/gitleaks/bad/gitleaks_bad_01.txt:aws-access-token:1
-/github/workspace/test/linters/gitleaks/bad/gitleaks_bad_01.txt:generic-api-key:2
-/tmp/lint/.github-personal-access-token:github-fine-grained-pat:1
```
[the remark this answers is not in the document]
Comment (ferrarimarco) [reply .gitleaksignore:5]: These are covered in the two different Gitleaks config files that we use for tests. We don't need to exclude the PAT file because we properly build the list of test cases.

File: docs/add-new-linter.md [hunk 2]
```diff
@@ -90,6 +90,7 @@ new tool, it should include:
 - Update the orchestration scripts to run the new tool:
 
   - `lib/linter.sh`
+  - `lib/linterCommands.sh`
```
[the remark this answers is not in the document]
Comment (ferrarimarco) [reply docs/add-new-linter.md:93]: Good catch! I'll fix this by amending the current commit.

**Comment by ferrarimarco** [comment]
@Hanse00 Here's a possible review path:

1. Review Dockerfile: install parallel
2. Review .gitignore: ignore terraform state stuff that we fetch before running tflint
3. Review Makefile, test/run-super-linter-tests.sh: move the logic to run linter tests from makefile to a script, use FILTER_REGEX_EXCLUDE to tell super-linter which test cases to run, instead of having special cases handled in buildFileList when test mode is on.
4. Checkov configuration files: Use Checkov config files to include the directories to lint so we don't need to account for special cases.
5. Gitleaks configuration files and workflows/ci.yml: remove .gitleaksignore and use a configuration file to ignore tests when linting the codebase, otherwise "bad" tests will fail validation
6. Review docs: Gitleaks now behaves as any other linter
7. Review linterCommands.sh, super_linter.rb, linter.sh (part): move commands from linter.sh to a dedicated file because we need to source this file from worker (we can't pass arrays using env variables)
8. Review validation.sh and log.sh: minor changes to export needed functions so subprocesses can find them. No need to define error count variables anymore because we use files to hold results and communicate across projects.
9. Review detectFiles.sh: minor changes, plus move all the "additional installation steps before running linters" in `RunAdditionalInstalls`. Before this change, it was scattered around several places.

At this point, you should be left with the three main files to review: linter.sh, buildFileList.sh, worker.sh

I'll follow up with another comment, so you can read this in the meantime

**Comment by ferrarimarco** [comment]
The main idea behind the changes in linter.sh, buildFileList.sh, worker.sh is that when there was a loop, we now have a `parallel` invocation that takes a certain number of arguments.

The main loops to consider are:

- buildFileList: Loop over each file in the codebase that we build with git or with find, and categorize each file according to its type. We refactored this loop so that we have multiple processes handling a portion of the list of files to categorize. This let us categorize multiple files in parallel.
- linter: Loop over the list of "languages" to lint. For each "language" to lint, we now have a `parallel` invocation of the `LintCodeBase` function that we define in worker.sh. This let us lint files for several languages in parallel.
- worker: Loop over the files to lint because many linters support linting more than one file for each invocation. This let us avoid paying the startup cost for these linters. In this case, there are a few corner cases to handle when linters aren't able to take more than file to lint at a time.

Finally, I had to refactor the return code logic a bit to handle three cases so we can properly run our test suite:

- All good -> return 0
- All linters returned errors -> return 1
- Some linters returned errors, but some returned ok -> return 2

**Review by Hanse00 — APPROVED** [review]
Appreciate the additional narration. I'm comfortable approving this.

**Comment by yermulnik** [comment]
ℹ️ JFYI for those who will stumble upon this PR wondering why `shfmt` started to lint shell files out of the blue since the `v6` of `super-linter`: inter alia this PR removed a historical workaround introduced back in `v3.13.0` to skip `shfmt` if there's no `.editorconfig` in `$GITHUB_WORKSPACE` (at that time `shfmt` probably required `.editorconfig` 🤔 I couldn't nail that 🤷🏻).

So since the `v6` release the `shfmt` isn't skipped anymore and can lint shell files as it is supposed to and, welp, now we have to align with shell formatting standards (my colleagues are starting to get confused with this new behavior 🤪)

Some refs:
* https://github.com/super-linter/super-linter/pull/642#issuecomment-683459600 + https://github.com/super-linter/super-linter/pull/642#issuecomment-683886242
* https://github.com/super-linter/super-linter/issues/2068
* https://github.com/super-linter/super-linter/blob/v3.13.0/lib/linter.sh#L1239-L1245
